File: src/routes/slow_request.rs
use std::path::Path;
use std::thread::sleep;
use std::time::Duration;

use crate::requests::{Request, Response, Status};

/// Process the `GET /slow_request`.
///
/// The function sleeps 5 secondes before returning the response, to simulate a slow
/// request to process.
///
/// # Returns
///
/// Returns the response to send with [`Response::send()`], with the HTML page
/// [`templates/slow_request.html`](/templates/slow_request.html).
///
/// # Panics
///
/// If the method [`Response::add_file()`] returns an error when adding a file.
///
/// # Examples
///
/// Check examples of [`WebServer::add_listener()`][add_listener],
/// to see how to add the function to process the `GET /slow_request`.
///
/// <!-- References -->
///
/// [add_listener]: crate::server::WebServer::add_listener()
pub fn get(request: Request) -> Response {
    let mut response = Response::from((request, Status::OK));
    response
        .add_file(Path::new("templates/slow_request.html"))
        .unwrap();

    sleep(Duration::from_secs(5));
    response
}

File: src/web-server/routes/slow_request.cpp
#include "slow_request.hpp"

#include "templates_directory.hpp"

#include <web-server/requests/request.hpp>
#include <web-server/requests/response.hpp>
#include <web-server/requests/status.hpp>

#include <chrono>
#include <thread>
#include <utility>

namespace {
    constexpr std::chrono::seconds SLEEP_TIME{5};
} // namespace

namespace web_server::routes::slow_request {
    using requests::Request;
    using requests::Response;
    using requests::Status;
    using tools::templates_directory;

    Response get(Request request) {
        auto response = Response::from(std::move(request), Status::Ok);
        response.add_file(templates_directory() / "slow_request.html");

        std::this_thread::sleep_for(SLEEP_TIME);

        return response;
    }
} // namespace web_server::routes::slow_request
